[MRG+1] FIX Hash collisions in the FeatureHasher

rth opened the pull request:
This PR fixes https://github.com/scikit-learn/scikit-learn/issues/7513 and extends the `non_negative` parameter in `FeatureHasher` and `HashingVectorizer` with a value of `'total'` which disables the hash collision handling (and makes the output more intuitive).

Here are [the annotations](http://r0h.eu/d/_hashing.html) for the compiled `_hashing.pyx`. Suggestions on how to better explain the `non_negative` behavior in the doc-strings are welcome.

The [documentation for the HashingVectorizer](http://scikit-learn.org/stable/modules/feature_extraction.html#feature-hashing) should also be improved IMO, e.g. to explain the preservation of the inner product in the hashed space, but maybe in a separate PR.

**Update:** this PR also supersedes https://github.com/scikit-learn/scikit-learn/pull/7502 that was a continuation of https://github.com/scikit-learn/scikit-learn/pull/5861 aiming to fix https://github.com/scikit-learn/scikit-learn/issues/3637 . It also resolves https://github.com/scikit-learn/scikit-learn/issues/2665 .

rth committed:
HashingVectorizer: optionaly disable alternate signs

rth committed:
HashingVectorizer updating documentation

rth committed:
Adding unit tests for non negative

jnothman commented:
Thanks for this. I'll try take a look tomorrow.

jnothman commented on sklearn/feature_extraction/_hashing.pyx:
```diff
@@ -63,7 +63,8 @@ def transform(raw_X, Py_ssize_t n_features, dtype):
             array.resize_smart(indices, len(indices) + 1)
             indices[len(indices) - 1] = abs(h) % n_features
-            value *= (h >= 0) * 2 - 1
+            if alternate_sign: # counter the effect of hash collision (issue #7513)
```
I don't think this description is correct.

jnothman replied:
(also, PEP8: keep two spaces before a comment)

jnothman commented on sklearn/feature_extraction/_hashing.pyx:
```diff
@@ -15,7 +15,7 @@ np.import_array()
 @cython.boundscheck(False)
 @cython.cdivision(True)
-def transform(raw_X, Py_ssize_t n_features, dtype):
+def transform(raw_X, Py_ssize_t n_features, dtype, char alternate_sign):
```
type should be `bint`, not `char`

jnothman commented on sklearn/feature_extraction/hashing.py:
```diff
@@ -53,11 +53,12 @@ class FeatureHasher(BaseEstimator, TransformerMixin):
-        When False, output values will have expected value zero.
+    non_negative : boolean or 'total', optional, default False
+        When True or False, an alternating sign is added to the counts as to
+        approximately conserve the inner product in the hashed space.
```
usually one would use "preserve" not "conserve"

jnothman commented on sklearn/feature_extraction/hashing.py:
```diff
@@ -53,11 +53,12 @@ class FeatureHasher(BaseEstimator, TransformerMixin):
-        When True, output values can be interpreted as frequencies.
-        When False, output values will have expected value zero.
+    non_negative : boolean or 'total', optional, default False
+        When True or False, an alternating sign is added to the counts as to
```
Not necessarily counts, here.

jnothman commented on sklearn/feature_extraction/hashing.py:
```diff
@@ -53,11 +53,12 @@ class FeatureHasher(BaseEstimator, TransformerMixin):
+        approximately conserve the inner product in the hashed space.
+        When True, an absolute value is additionally applied to the result
+        prior to returning it.
+        When 'total' all counts are positive which disables collision handling.
```
It doesn't really disable collision handling, does it? It just does not faithfully preserve inner product. The output, however, can still include negative values if input features had negative values. So `total` might not be correct nomenclature either. I think you might say "this is suitable for non-negative features; each feature value is then the sum of all features with that hash"

jnothman commented on sklearn/feature_extraction/hashing.py:
```diff
@@ -148,6 +152,6 @@ def transform(self, raw_X, y=None):
                           shape=(n_samples, self.n_features))
         X.sum_duplicates()  # also sorts the indices
-        if self.non_negative:
+        if self.non_negative is True:  # if non_negative == 'total', X>0 anyway
```
Not necessarily true. We need to decide how to handle that case.

One solution is to add a further parameter... But honestly, if there is negative input, `non_negative` should be `False`.

jnothman commented on sklearn/feature_extraction/tests/test_feature_hasher.py:
```diff
@@ -107,3 +107,21 @@ def test_hasher_zeros():
+def test_hasher_non_negative():
+    raw_X = [["foo", "bar", "baz"]]
+
+    def it():  # iterable
```
why?

jnothman replied:
why not use `raw_X` directly?

And convention would use `X` for input, and `Xt` for output.

jnothman reviewed (changes requested):
I'm coming to consider the idea that the current `non_negative` is rubbish, and that maybe we should deprecate it an install another parameter in its place...

rth commented:
@jnothman Thanks a lot for the detailed review! I wasn't sure about boolean in Cython and  did miss the second space before the comment in the `.pyx`. ) I agree that having 3 values for the current `non_negative`  parameter not great and that something should be done about it.

For the formulation, I still think that saying that this mechanism would allow to "approximately preserve inner product in the hashed space" is misleading. As is illustrated in [my other comment](https://github.com/scikit-learn/scikit-learn/issues/7513#issuecomment-252322348), a more exact formulation would be that it allows to better conserve the inner product, for a) small hashing table sizes  b) when binary counts are not used, so at the end it is still about how collisions are handled with respect to this property.

I'll look into addressing the other comments in your review soon.

rth committed:
Addressing review comments

rth commented:
Sorry for the late response. @jnothman  I addressed review comments above, and the ones in [your post](https://github.com/scikit-learn/scikit-learn/issues/7513#issuecomment-252418149) in the original issue. 

I'm just not sure about when the deprecation warning should be raised: in the current PR it is raised both for `non_negative=True` (to warn that it's deprecated) and for  `non_negative='total'` (to warn that this option will be renamed to `non_negative=True` ), but that doesn't sound right as this means that the users won't be able to do anything to make this warning go away for 2 versions. Should I deprecate just `non_negative=True` then?

jnothman commented:
Yes, don't deprecate `non_negative='total'`.

jnothman commented on sklearn/feature_extraction/hashing.py:
```diff
@@ -79,6 +85,14 @@ class FeatureHasher(BaseEstimator, TransformerMixin):
+                             " True, False, 'total'.")
+        if non_negative in ['total', True]:
+            warnings.warn("the option non_negative=True has been deprecated"
+                          " in 0.19. As of 0.21 non_negative='total' would be"
```
Perhaps "From version 0.21, non_negative=True will be interpreted as non_negative='total'."

jnothman commented on sklearn/feature_extraction/tests/test_feature_hasher.py:
```diff
@@ -107,3 +107,21 @@ def test_hasher_zeros():
+
+    X = FeatureHasher(non_negative=False,
+                      input_type='string').fit_transform(it())
+    assert_true((X.data > 0).any() and (X.data < 0).any())
```
surely this test is more useful if we check that there's a zero?

rth replied:
Well the point here was to test that the output has both negative and positive values, not that is has zeros (which it doesn't in this case). Changed the test to be more verbose about it.

rth committed:
Adressing more review comments

rth commented:
Thanks! I made the changes requested above.

The only comment I'm not sure how to address is about when [the absolute value should be applied](https://github.com/scikit-learn/scikit-learn/pull/7565#discussion_r81959884). You mean that if the features counts are negative to begin with, then with this code `self.non_negative==True` will make the output positive, but `self.non_negative=='total'` will not? But at the same time unnecessarily applying the absolute value to feature counts with HashingVectorizer (probably 95% use cases for this) would have some overhead, wouldn't it? Otherwise the solution could be to rename the `non_negative` parameter to something else (as with this PR making the output positive/negative would be just a side effect of the improved inner product preservation, which is what this parameter does in the end), but that would give additional head ache with backward compatibility...

rth committed:
Missed a few review comments

jnothman commented on sklearn/feature_extraction/tests/test_feature_hasher.py:
```diff
@@ -107,3 +107,18 @@ def test_hasher_zeros():
+    assert_true((Xt.data >= 0).all())  # zeros are acceptable
+    Xt = FeatureHasher(non_negative='total',
+                      input_type='string').fit_transform(X)
+    assert_true((Xt.data > 0).all())  # strictly positive counts
```
But this isn't testing that `non_negative='total'` is working, is it?

rth replied:
You are right, thanks, changed the test to actually have a hash collision so this part is tested.

jnothman commented:
Could you run a benchmark on the overhead of `abs`? otherwise maybe we could indeed introduce a parameter `sign_flipping=True` (or something) and deprecate `non_negative`.

rth committed:
Adding an example that produces a hash collision

rth commented:
I checked the overhead of abs on the 20 newsgroups when using HashingVectorizer, is is negligible (1 ms out of a 3.65s total run time). Another issue is that right now there is no way of doing just the plain feature hashing (i.e. just cont the occurrences in each element of the hashing table):   if `non_negative = False` this ~ alternating sign is applied, if `non_negative in [True, 'total'] we also apply the absolute value. So in the case when some features are negative initially, there is no way to just count them without an additional abs. This is because the current `non_negative` parameter mixes abs with the "alternating sign" functionality.

I agree that adding a new parameter (how about `rand_sparse_proj` or `altern_sign`?) and deprecating `non_negative` would be annoying, but it may be the best thing to here?

jnothman commented:
I think `altern_sign=True` would be clear enough. Let's go with that..?

jnothman commented:
Make it `alternate_sign`

rth committed:
Adding a alternate_sign parameter to the FeatureHasher

jnothman commented on sklearn/feature_extraction/hashing.py:
```diff
@@ -53,11 +54,16 @@ class FeatureHasher(BaseEstimator, TransformerMixin):
+        When True, an absolute value is applied to the features matrix prior to
+        returning it. When used in conjunction with alternate_sign=True, this
+        significantly reduces the inner product preservation property.
+        This option is deprecated as of 0.19.
```
you can use the `.. deprecated` directive. Be sure to specify when it'll be removed (0.21).

jnothman replied:
Actually, not so important seeing as you have 0.21 in the code

jnothman commented:
This looks good. But in case someone were to reimplement it with an abs, please add a test with mixed-sign data.

rth committed:
Adding a mixed-sign test and deprecated directives

rth commented:
@jnothman Thanks for the feedback! I added a test with mixed-sign data and the `..deprecated` directives.

jnothman commented:
LGTM. Thanks very much!

jnothman commented:
Please add an entry in whats_new

rth committed:
Added an entry in what's new

rth commented:
Sorry for the delay, I rebased and added an entry in whats_new. Would someone be able to do a second review of this, PR please? cc @raghavrv 
This PR is still marked by github as having requested changes by @jnothman while it was approved above (well apart for the "what's new" section). Thanks!

rth commented:
Also when I go to https://github.com/scikit-learn/scikit-learn/projects/5 for 0.19 both parent issues that this PR aims to address are marked as "Issues without PR", a [stalled PR](https://github.com/scikit-learn/scikit-learn/pull/7502)  that was continued in PR https://github.com/scikit-learn/scikit-learn/pull/7502, itself superseded by this PR is marked as "PRs in progress/stalled" and this PR is not in the list of "PRs that need reviews" ...

jnothman commented:
I think you can that Project board for nothing :)

On 24 January 2017 at 22:52, Roman Yurchak <notifications@github.com> wrote:

> Also when I go to https://github.com/scikit-learn/scikit-learn/projects/5
> for 0.19 both parent issues that this PR aims to address are marked as
> "Issues without PR", a stalled PR
> <https://github.com/scikit-learn/scikit-learn/pull/7502> that was
> continued in PR #7502
> <https://github.com/scikit-learn/scikit-learn/pull/7502>, itself
> superseded by this PR is marked as "PRs in progress/stalled" and this PR is
> not in the list of "PRs that need reviews" ...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7565#issuecomment-274783509>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz63JaR5wlFusJRjAO1rMkT5wUZhyJks5rVeXggaJpZM4KM2mB>
> .
>

jnothman commented:
Or perhaps not. Looks like someone's been putting things in there, but I'm
not able to tell who, or when, because...  well it's github but there's no
VCS.

On 24 January 2017 at 22:58, Joel Nothman <joel.nothman@gmail.com> wrote:

> I think you can that Project board for nothing :)
>
> On 24 January 2017 at 22:52, Roman Yurchak <notifications@github.com>
> wrote:
>
>> Also when I go to https://github.com/scikit-learn/scikit-learn/projects/5
>> for 0.19 both parent issues that this PR aims to address are marked as
>> "Issues without PR", a stalled PR
>> <https://github.com/scikit-learn/scikit-learn/pull/7502> that was
>> continued in PR #7502
>> <https://github.com/scikit-learn/scikit-learn/pull/7502>, itself
>> superseded by this PR is marked as "PRs in progress/stalled" and this PR is
>> not in the list of "PRs that need reviews" ...
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/scikit-learn/scikit-learn/pull/7565#issuecomment-274783509>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAEz63JaR5wlFusJRjAO1rMkT5wUZhyJks5rVeXggaJpZM4KM2mB>
>> .
>>
>
>

raghavrv commented on doc/whats_new.rst:
```diff
@@ -198,6 +198,12 @@ Bug fixes
+     applied a sparse random projection to the hashed features, preventing
+     the use of `sklearn.feature_extraction.text.HashingVectorizer` in a
+     pipeline with  `sklearn.feature_extraction.text.TfidfTransformer`.
+     :issue:`3637`, :issue:`7513` by `Roman Yurchal <rth>`.
```
You can reference the PR alone... (#7513)

raghavrv commented on sklearn/feature_extraction/_hashing.pyx:
```diff
@@ -63,7 +63,9 @@ def transform(raw_X, Py_ssize_t n_features, dtype):
             array.resize_smart(indices, len(indices) + 1)
             indices[len(indices) - 1] = abs(h) % n_features
-            value *= (h >= 0) * 2 - 1
+            if alternate_sign:  # improve inner product preservation
```
Style nitpick: Could you move the comment to above the `if ...`?

raghavrv commented on sklearn/feature_extraction/_hashing.pyx:
```diff
@@ -63,7 +63,9 @@ def transform(raw_X, Py_ssize_t n_features, dtype):
             indices[len(indices) - 1] = abs(h) % n_features
-            value *= (h >= 0) * 2 - 1
+            if alternate_sign:  # improve inner product preservation
+                                # in the hashed space (issue #7513)
```
Also please remove the issue reference

raghavrv reviewed (changes requested):
I'm not very familiar with this part of the codebase... but this does fix the issue referenced.

Also do you want a test with `HashingVectorizer` and `TfidfTransformer` in a pipeline?

rth committed:
Addressing review comments

rth commented:
Thanks for the review, @raghavrv . I addressed your comments. 

Well, adding a test with `HashingVectorizer` and `TfidfTransformer` in a pipeline would be more like an integration test; we know that issue https://github.com/scikit-learn/scikit-learn/issues/3637 is caused by the `HashingVectorizer` producing zeros in returned document-term matrix `X` (due to hash collisions), which happens in the `FeatureHasher`. I did do initially tests on the `HashingVectorizer`, however @jnothman [suggested](https://github.com/scikit-learn/scikit-learn/pull/7502#issuecomment-250029829) that I test `FeatureHasher` instead. Currently we test that were [previously we had a hash collision producing a zero](https://github.com/scikit-learn/scikit-learn/pull/7565/files#diff-2d64900f0cd3b75552d31482cee9b0e6R120) after this PR all elements  [are strictly positive](https://github.com/scikit-learn/scikit-learn/pull/7565/files#diff-2d64900f0cd3b75552d31482cee9b0e6R128) which should cover the pipelining issue. But I can add another integration test if you think I should?

Also Circle CI is currently failing with the following error,
```
+ ./miniconda.sh -b -p
./miniconda.sh: option requires an argument -- p
Error: did not recognize option, please try -h
```
not sure what this is due to.

lesteve commented:
> Also Circle CI is currently failing with the following error,

Try rebasing on master. I am not 100% sure but I would guess that CircleCI uses the circle.yml from your fork that does not have MINICONDA_PATH setup. Then we have a tweak to run CircleCI on the result of the merge of the PR into master and build_doc.sh uses MINICONDA_PATH which is unset.

rth commented:
Thanks @lesteve , rebasing fixed the Circle CI issue.

lesteve commented:
> Thanks @lesteve , rebasing fixed the Circle CI issue.

Great, I believe other PRs are affected so this is good to make sure this fixes the problem.

rth committed:
Merge branch 'master' into hashing_vect-nn

rth commented:
> Thanks for the review, @raghavrv . I addressed your comments.
@raghavrv  Please let me know if you have any other comments. Thanks..

raghavrv commented:
Sorry for the delay in response... This needs a merge with master. Could you take care of it? I'll try to review soon...

rth committed:
Merge branch 'master' into hashing_vect-nn

rth commented:
Thanks @raghavrv ! I fixed the conflict...

lesteve commented on doc/whats_new.rst:
```diff
@@ -239,9 +239,17 @@ Bug fixes
      :issue:`8160` by :user:`CJ Carey <perimosocordiae>`.
 
+
+   - Fix a bug where `sklearn.feature_extraction.FeatureHasher` mandatorily
```
This is rst not md, so you want to use double-backticks. Best is to use the sphinx directives `:class:` or `:func:`. Look around in this file for examples.

Note this happens a few times below as well.

rth committed:
Fixed rst formatting

rth commented:
Thanks for the comment @lesteve , fixed the rst formatting...

jnothman commented:
@raghavrv and @lesteve, you both made requests here. Are we ready to merge?

lesteve commented:
I am not very familiar with text so I am afraid I am not the best one to review this one.

raghavrv commented:
@jnothman do you think it's better to have `random` somewhere in the parameter name just to make things clear that it does not always alternate...

Maybe `randomly_alternate_sign` or (`random_alter_sign` if you prefer something short)?

raghavrv commented on sklearn/feature_extraction/tests/test_feature_hasher.py:
```diff
@@ -106,3 +106,44 @@ def test_hasher_zeros():
+
+def test_hasher_alternate_sign():
+    X = [["foo", "bar", "baz", "investigation need", "records"]]
+    # the last two tokens produce a hash collision that sums as 0
```
Can you move this comment above the previous line?

(+1 for a well thought of test case. Thx)

jnothman commented:
It's not random, it's a function of the hash...

raghavrv commented on sklearn/feature_extraction/tests/test_feature_hasher.py:
```diff
@@ -106,3 +106,44 @@ def test_hasher_zeros():
+    assert_true(len(Xt.data) < len(X[0]))
+    assert_true((Xt.data == 0.).any())
+
+    Xt = FeatureHasher(alternate_sign=True, non_negative=True,
```
This will raise a deprecation warning right? can you wrap it using `ignore_warnings(DeprecatedWarning)` maybe?

raghavrv commented on sklearn/feature_extraction/text.py:
```diff
@@ -405,11 +404,18 @@ class HashingVectorizer(BaseEstimator, VectorizerMixin):
+    alternate_sign : boolean, optional, default True
+        When True, an alternating sign is added to the features as to
+        approximately conserve the inner product in the hashed space even for
+        small n_features. This approach is similar to sparse random projection.
```
you need a new feature (or feature added?) tag?

raghavrv replied:
(has been addressed below)

raghavrv reviewed (changes requested):
Thanks for the PR. Looks good to me, pending minor comments... I'll make a final pass and merge once you address those...

raghavrv commented:
> It's not random, it's a function of the hash...

Ah! got it thanks :)

rth committed:
Merge branch 'master' into hashing_vect-nn

rth committed:
Fixing review comments

rth committed:
Merge branch 'hashing_vect-nn' of github.com:rth/scikit-learn into hashing_vect-nn

rth commented:
Thanks for the review @raghavrv ! I addressed all your comments (I think). Appveyor would take a day or so to build, unfortunately..

raghavrv commented:
Thanks @rth. I'll approve it for now... Please ping me when appveyor passes. I'll merge it... (Or is it okay to merge without appveyor, @jnothman?)

jnothman commented:
By induction, appveyor will provably pass :p

Thanks @rth!
